go/src/pluginmgr/pluginmgr_common/pluginmgr_helper.go

```diff
@@ -0,0 +1,496 @@
+	}
+
+	return nil
+}
+
+// DeRegisterActionWithEngine : DeRegister plugin with engine
+func (plmgr *PluginManager) deRegisterActionWithEngine(pluginName string) error {
+	if err := plmgr.clientTx.DeregisterAction(pluginName); err != nil {
+		//delete(plmgr.plugins, pluginName) // TODO: Goutham : , Is this needed instead. Can't we store failed plugins ?
+		//delete(plmgr.pluginMetadata, pluginName) // TODO: Goutham : verify Is this needed instead. Can't we store failed plugins ?
+		return lomcommon.LogError("Failed to register plugin %s with engine", pluginName)
+	}
+
+	//delete(plmgr.plugins, pluginName) // TODO: Goutham : verify, Is this needed instead.
+	//delete(plmgr.pluginMetadata, pluginName) // TODO: Goutham : verify Is this needed instead.
+	return nil
+}
+
+// TODO: Goutham : Add this function in lomcommon framework
+// If requirement is to add periodic log entry with shorttime and then falback to longtime, use this function.
+func (plmgr *PluginManager) AddPeriodicLogWithTimeouts(ID string, message string, shortTimeout time.Duration,
+	longTimeout time.Duration, context string) chan bool {
+	// Create a channel to listen for stop signals to kill timer
+	stopchannel := make(chan bool)
+
+	lomcommon.GetGoroutineTracker().Start("AddPeriodicLogWithTimeouts"+ID+lomcommon.GetUUID(), func() {
+		// First add periodic log witj short timeout
+		lomcommon.AddPeriodicLogInfo(ID, message, int(shortTimeout.Seconds())) // call lomcommon framework
+
+		// Wait for the short timeout to expire or for stop signal
+		select {
+		case <-time.After(shortTimeout):
+			// after short timeout expiry, update timer to longtimeout
+			lomcommon.UpdatePeriodicLogTime(ID, int(longTimeout.Seconds()))
+			break
+		case <-stopchannel:
+			lomcommon.RemovePeriodicLogEntry(ID)
+			return
+		}
+
+		// Wait for the stop signal
+		<-stopchannel
+
+		// Stop signal received, remove the periodic log entry
+		lomcommon.RemovePeriodicLogEntry(ID)
+	})
+
+	return stopchannel
+}
+
+/*****************************************************************************************/
+/* 	  Helper Functions										      			             */
+/*****************************************************************************************/
+
+/*
+ *  Create Plugin Instance. Multiple plugins can be created from this function
+ * When a new plugin is added, add a case here to create the plugin instance and define the plugin in plugins_common package
+ *
+ * Input:
+ *  pluginID - Plugin ID(pluginname, version) from plugins_common package
+ *  pluginData - Plugin Data passed to plugin. For e.g. heartbeats timer, etc.
+ *
+ * Output:
+ *  none -
+ *
+ * Return:
+ *  plugin - Plugin instance
+ *  pluginmetadata - Plugin metadata
+ *  error - Error if any
+ */
+//TODO : Goutham : Look into beetter way to do this.
+func CreatePluginInstance(pluginID plugins_common.PluginId, actionCfg *lomcommon.ActionCfg_t) (plugins_common.Plugin, plugins_common.IPluginMetadata, error) {
```

Comment:
Here is an idea:
Each Plugin object adds an init func and add its object constructor function in a list like below.
Being init, as all init functions are called there is no concurrency issue,

type PluginConstructor func(...)plugins_common.Plugin
map[action-name] PluginConstructor

Your loading API uses it to call appropriate fn to construct.

Reply:
Its great idea. 
I was working on other idea based on Reflect which don't need to have any init() . I will weight pros and cons of both and add this in next  PR. 

// Get the type of the plugin based on its name
	name := "LinkFlapPluginDetection"
	pluginType := reflect.TypeOf(&plugins_files[name]{})
	if pluginType == nil {
		return nil, nil, lomcommon.LogError("plugin not found: %s", pluginID.Name)
	}

	// Create an instance of the plugin using reflection
	pluginValue := reflect.New(pluginType).Elem()
	plugin := pluginValue.Addr().Interface().(plugins_common.Plugin)